### PR TITLE
OpenAPI Extension Annotations 

### PR DESCRIPTION
**Pull request by RoshniNaveenaS**
In this PR , we will be supporting `@OpenAPI.Extensions` in three levels:

1. Schema Level
2. Root level
3. Operation level ( action or function)

Example:
In CDS:

```
service SampleService {
    annotate Employee with @(
    OpenAPI.Extensions:{
       abc : {
        g:'h',
        q:'p',
        dj : {
            q:'fk'
        }
       },
       bn:{
        c :'k'
       }
    }
);
    entity Employee {
        key ID        : Integer;
            firstName : String;
            lastName  : String;
            age       : Integer;
    }

    annotate promoteEmployee with @(
    OpenAPI.Extensions:{
       action : {
        g:'h',
        q:'p',
        dj : {
            q:'fk'
        }
       },
       bn:{
        c :'k'
       }
    });
    action promoteEmployee(employeeID: Integer) returns String;

    annotate getEmployeeAge with @(
    OpenAPI.Extensions:{
       function : {
        g:'h',
        q:'p',
        dj : {
            q:'fk'
        }
       },
       bn:{
        c :'k'
       }
    }
);
    function getEmployeeAge(employeeID: Integer) returns Integer;
}


annotate SampleService with @(
    OpenAPI.Extensions:{
       abc : {
        g:'h',
        q:'p',
        dj : {
            q:'fk'
        }
       },
       bn:{
        c :'k'
       }
    }
);
```

In openAPI:

```
{
  "openapi": "3.0.2",
  "info": {
    "title": "Use @title: '...' on your CDS service to provide a meaningful title.",
    "description": "Use @Core.LongDescription: '...' on your CDS service to provide a meaningful description.",
    "version": ""
  },
  "x-sap-api-type": "ODATAV4",
  "x-odata-version": "4.01",
  "x-sap-shortText": "Use @Core.Description: '...' on your CDS service to provide a meaningful short text.",
  "servers": [
    {
      "url": "/rest/sample"
    }
  ],
  "tags": [
    {
      "name": "Employee"
    }
  ],
  "paths": {
    "/$batch": {
      "post": {
        "summary": "Sends a group of requests",
        "description": "Group multiple requests into a single request payload, see [Batch Requests](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_BatchRequests).\n\n*Please note that \"Try it out\" is not supported for this request.*",
        "tags": [
          "Batch Requests"
        ],
        "requestBody": {
          "required": true,
          "description": "Batch request",
          "content": {
            "multipart/mixed;boundary=request-separator": {
              "schema": {
                "type": "string"
              },
              "example": "--request-separator\nContent-Type: application/http\nContent-Transfer-Encoding: binary\n\nGET Employee HTTP/1.1\nAccept: application/json\n\n\n--request-separator--"
            }
          }
        },
        "responses": {
          "200": {
            "description": "Batch response",
            "content": {
              "multipart/mixed": {
                "schema": {
                  "type": "string"
                },
                "example": "--response-separator\nContent-Type: application/http\n\nHTTP/1.1 200 OK\nContent-Type: application/json\n\n{...}\n--response-separator--"
              }
            }
          },
          "4XX": {
            "$ref": "#/components/responses/error"
          }
        }
      }
    },
    "/Employee": {
      "get": {
        "summary": "Retrieves a list of employee.",
        "tags": [
          "Employee"
        ],
        "parameters": [
          {
            "$ref": "#/components/parameters/top"
          },
          {
            "$ref": "#/components/parameters/skip"
          },
          {
            "$ref": "#/components/parameters/search"
          },
          {
            "name": "$filter",
            "description": "Filter items by property values, see [Filtering](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionfilter)",
            "in": "query",
            "schema": {
              "type": "string"
            }
          },
          {
            "$ref": "#/components/parameters/count"
          },
          {
            "name": "$orderby",
            "description": "Order items by property values, see [Sorting](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionorderby)",
            "in": "query",
            "explode": false,
            "schema": {
              "type": "array",
              "uniqueItems": true,
              "items": {
                "type": "string",
                "enum": [
                  "ID",
                  "ID desc",
                  "firstName",
                  "firstName desc",
                  "lastName",
                  "lastName desc",
                  "age",
                  "age desc"
                ]
              }
            }
          },
          {
            "name": "$select",
            "description": "Select properties to be returned, see [Select](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionselect)",
            "in": "query",
            "explode": false,
            "schema": {
              "type": "array",
              "uniqueItems": true,
              "items": {
                "type": "string",
                "enum": [
                  "ID",
                  "firstName",
                  "lastName",
                  "age"
                ]
              }
            }
          }
        ],
        "responses": {
          "200": {
            "description": "Retrieved employee",
            "content": {
              "application/json": {
                "schema": {
                  "type": "object",
                  "title": "Collection of Employee",
                  "properties": {
                    "@count": {
                      "$ref": "#/components/schemas/count"
                    },
                    "value": {
                      "type": "array",
                      "items": {
                        "$ref": "#/components/schemas/SampleService.Employee"
                      }
                    }
                  }
                }
              }
            }
          },
          "4XX": {
            "$ref": "#/components/responses/error"
          }
        }
      },
      "post": {
        "summary": "Creates a single employee.",
        "tags": [
          "Employee"
        ],
        "requestBody": {
          "description": "New employee",
          "required": true,
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/SampleService.Employee-create"
              }
            }
          }
        },
        "responses": {
          "201": {
            "description": "Created employee",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/SampleService.Employee"
                }
              }
            }
          },
          "4XX": {
            "$ref": "#/components/responses/error"
          }
        }
      }
    },
    "/Employee({ID})": {
      "parameters": [
        {
          "description": "key: ID",
          "in": "path",
          "name": "ID",
          "required": true,
          "schema": {
            "type": "integer",
            "format": "int32"
          }
        }
      ],
      "get": {
        "summary": "Retrieves a single employee.",
        "tags": [
          "Employee"
        ],
        "parameters": [
          {
            "name": "$select",
            "description": "Select properties to be returned, see [Select](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionselect)",
            "in": "query",
            "explode": false,
            "schema": {
              "type": "array",
              "uniqueItems": true,
              "items": {
                "type": "string",
                "enum": [
                  "ID",
                  "firstName",
                  "lastName",
                  "age"
                ]
              }
            }
          }
        ],
        "responses": {
          "200": {
            "description": "Retrieved employee",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/SampleService.Employee"
                }
              }
            }
          },
          "4XX": {
            "$ref": "#/components/responses/error"
          }
        }
      },
      "patch": {
        "summary": "Changes a single employee.",
        "tags": [
          "Employee"
        ],
        "requestBody": {
          "description": "New property values",
          "required": true,
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/SampleService.Employee-update"
              }
            }
          }
        },
        "responses": {
          "204": {
            "description": "Success"
          },
          "4XX": {
            "$ref": "#/components/responses/error"
          }
        }
      },
      "delete": {
        "summary": "Deletes a single employee.",
        "tags": [
          "Employee"
        ],
        "responses": {
          "204": {
            "description": "Success"
          },
          "4XX": {
            "$ref": "#/components/responses/error"
          }
        }
      }
    },
    "/getEmployeeAge": {
      "get": {
        "summary": "Invokes function getEmployeeAge",
        "tags": [
          "Service Operations"
        ],
        "parameters": [
          {
            "required": true,
            "in": "query",
            "name": "employeeID",
            "schema": {
              "type": "integer",
              "format": "int32",
              "nullable": true,
              "default": "null"
            }
          }
        ],
        "responses": {
          "200": {
            "description": "Success",
            "content": {
              "application/json": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "value": {
                      "type": "integer",
                      "format": "int32",
                      "nullable": true
                    }
                  }
                }
              }
            }
          },
          "4XX": {
            "$ref": "#/components/responses/error"
          }
        },
        "x-sap-function": {
          "g": "h",
          "q": "p",
          "dj": {
            "q": "fk"
          }
        },
        "x-sap-bn": {
          "c": "k"
        }
      }
    },
    "/promoteEmployee": {
      "post": {
        "summary": "Invokes action promoteEmployee",
        "tags": [
          "Service Operations"
        ],
        "responses": {
          "200": {
            "description": "Success",
            "content": {
              "application/json": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "value": {
                      "type": "string",
                      "nullable": true
                    }
                  }
                }
              }
            }
          },
          "4XX": {
            "$ref": "#/components/responses/error"
          }
        },
        "x-sap-action": {
          "g": "h",
          "q": "p",
          "dj": {
            "q": "fk"
          }
        },
        "x-sap-bn": {
          "c": "k"
        },
        "requestBody": {
          "description": "Action parameters",
          "content": {
            "application/json": {
              "schema": {
                "type": "object",
                "properties": {
                  "employeeID": {
                    "type": "integer",
                    "format": "int32",
                    "nullable": true
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "SampleService.Employee": {
        "title": "Employee",
        "type": "object",
        "properties": {
          "ID": {
            "type": "integer",
            "format": "int32"
          },
          "firstName": {
            "type": "string",
            "nullable": true
          },
          "lastName": {
            "type": "string",
            "nullable": true
          },
          "age": {
            "type": "integer",
            "format": "int32",
            "nullable": true
          }
        },
        "x-sap-abc": {
          "g": "h",
          "q": "p",
          "dj": {
            "q": "fk"
          }
        },
        "x-sap-bn": {
          "c": "k"
        }
      },
      "SampleService.Employee-create": {
        "title": "Employee (for create)",
        "type": "object",
        "properties": {
          "ID": {
            "type": "integer",
            "format": "int32"
          },
          "firstName": {
            "type": "string",
            "nullable": true
          },
          "lastName": {
            "type": "string",
            "nullable": true
          },
          "age": {
            "type": "integer",
            "format": "int32",
            "nullable": true
          }
        },
        "required": [
          "ID"
        ]
      },
      "SampleService.Employee-update": {
        "title": "Employee (for update)",
        "type": "object",
        "properties": {
          "firstName": {
            "type": "string",
            "nullable": true
          },
          "lastName": {
            "type": "string",
            "nullable": true
          },
          "age": {
            "type": "integer",
            "format": "int32",
            "nullable": true
          }
        }
      },
      "count": {
        "anyOf": [
          {
            "type": "number"
          },
          {
            "type": "string"
          }
        ],
        "description": "The number of entities in the collection. Available when using the [$count](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptioncount) query option."
      },
      "error": {
        "type": "object",
        "required": [
          "error"
        ],
        "properties": {
          "error": {
            "type": "object",
            "required": [
              "code",
              "message"
            ],
            "properties": {
              "code": {
                "type": "string"
              },
              "message": {
                "type": "string"
              },
              "target": {
                "type": "string"
              },
              "details": {
                "type": "array",
                "items": {
                  "type": "object",
                  "required": [
                    "code",
                    "message"
                  ],
                  "properties": {
                    "code": {
                      "type": "string"
                    },
                    "message": {
                      "type": "string"
                    },
                    "target": {
                      "type": "string"
                    }
                  }
                }
              },
              "innererror": {
                "type": "object",
                "description": "The structure of this object is service-specific"
              }
            }
          }
        }
      }
    },
    "parameters": {
      "top": {
        "name": "$top",
        "in": "query",
        "description": "Show only the first n items, see [Paging - Top](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptiontop)",
        "schema": {
          "type": "integer",
          "minimum": 0
        },
        "example": 50
      },
      "skip": {
        "name": "$skip",
        "in": "query",
        "description": "Skip the first n items, see [Paging - Skip](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionskip)",
        "schema": {
          "type": "integer",
          "minimum": 0
        }
      },
      "count": {
        "name": "$count",
        "in": "query",
        "description": "Include count of items, see [Count](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptioncount)",
        "schema": {
          "type": "boolean"
        }
      },
      "search": {
        "name": "$search",
        "in": "query",
        "description": "Search items by search phrases, see [Searching](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionsearch)",
        "schema": {
          "type": "string"
        }
      }
    },
    "responses": {
      "error": {
        "description": "Error",
        "content": {
          "application/json": {
            "schema": {
              "$ref": "#/components/schemas/error"
            }
          }
        }
      }
    }
  },
  "x-sap-abc": {
    "g": "h",
    "q": "p",
    "dj": {
      "q": "fk"
    }
  },
  "x-sap-bn": {
    "c": "k"
  }
}
```

### CODE REVIEW

File: test/lib/compile/openapi.test.js [hunk 1]
```diff
@@ -391,4 +391,32 @@ describe('OpenAPI export', () => {
         expect(openAPI.externalDocs.url).toBe('https://help.sap.com/docs/product/123.html');
   }
   );
+
+  test('OpenAPI annotations: @OpenAPI.Extensions annotation is added to the openapi document', () => {
+    const csn = cds.compile.to.csn(`
+      namespace sap.OpenAPI.test;
+      @OpenAPI.Extensions: {
```
Comment (jeevitha011) [inline test/lib/compile/openapi.test.js:398]: Additional cases for test: 

- nested annotation
- extension annotation key to start with `sap` or  `x-sap`(edge case)

For reference: https://github.com/cap-js/asyncapi/blob/main/test/lib/compile/input/valid/base.cds

**Review by jeevitha011 — CHANGES_REQUESTED** [review]

File: test/lib/compile/openapi.test.js (hunk 1 above)
Comment (RoshniNaveenaS) [reply test/lib/compile/openapi.test.js:398]: Have used the same test case to civer both these scenarios.

File: test/lib/compile/openapi.test.js [hunk 2]
```diff
@@ -424,4 +424,37 @@ service CatalogService {
...
+      }
+        function F1(param: String) returns String;
+          
+          }`);
+    const openAPI = toOpenApi(csn);
+    expect(openAPI).toBeDefined();
+    expect(openAPI['x-sap-compliance-level']).toBe('sap:base:v1');
+    expect(openAPI.components.schemas["sap.OpenAPI.test.A.E1"]["x-sap-dpp-is-potentially-sensitive"]).toBe('true');
+    expect(openAPI.paths["/F1"].get["x-sap-operation-intent"]).toBe('read-collection');
+    expect(openAPI.paths["/F1"].get["x-sap-deprecated-operation"].deprecationDate).toBe('2022-12-31');
```
Comment (jeevitha011) [inline test/lib/compile/openapi.test.js:458]: check can be added for nested annotation (successorOperationId) as well

File: lib/compile/csdl2openapi.js [hunk 3]
```diff
@@ -106,7 +106,60 @@ module.exports.csdl2openapi = function (
...
+        containerSchema = csdl.$EntityContainer ? csdl[namespace] : {};
+        }
+        else if(level === 'schema' || level === 'operation'){
+            containerSchema = csdl;
+        }
+
+        for (const [key, value] of Object.entries(containerSchema)) {
+            if (key.startsWith('@OpenAPI.Extensions')) {
+                const annotationProperties = key.split('@OpenAPI.Extensions.')[1];
+                const keys = annotationProperties.split('.');
```
Comment (jeevitha011) [inline lib/compile/csdl2openapi.js:129]: Processing for fields in `Defined Specification Extensions` like `x-sap-stateinfo` needs to be implemented.

File: test/lib/compile/openapi.test.js (hunk 2 above)
Comment (RoshniNaveenaS) [reply test/lib/compile/openapi.test.js:458]: made changes

File: test/lib/compile/openapi.test.js [hunk 4]
```diff
@@ -424,4 +424,47 @@ service CatalogService {
         expect(openAPI.externalDocs.url).toBe('https://help.sap.com/docs/product/123.html');
   }
   );
+
+  test('OpenAPI annotations: @OpenAPI.Extensions annotation is added to the openapi document', () => {
+    const csn = cds.compile.to.csn(`
+      namespace sap.OpenAPI.test;
+      @OpenAPI.Extensions: {
+        ![compliance-level]: 'sap:base:v1',
+        ![x-sap-ext-overview]: {
```
Comment (jeevitha011) [inline test/lib/compile/openapi.test.js:433]: shouldn't we add an `expect-toBe` check for `x-sap-ext-overview` nested value also?

**Review by jeevitha011 — CHANGES_REQUESTED** [review]

**Review by jeevitha011 — APPROVED** [review]
lgtm
